Documentation for Unstructured QA (#55)
cisco/mindmeld: source/userguide/kb.rst

.. note::

   We can set the ``size`` parameter of the :meth:`execute()` method to specify the maximum number of records.

Dealing with unstructured text
------------------------------

> The knowledge bases described in the previous sections are comprised of clearly defined data types whose pattern makes them easily searchable. We call them structured data.

Reviewer: ```suggestion
The knowledge bases described in the previous sections are comprised of clearly defined data types whose format makes them easily searchable. We call them structured data.
```

Reviewer: Wanted to find a better word than "pattern".

Alternately, "The knowledge bases described in the previous sections are comprised of clearly defined data types that are easily searchable. We call them structured data."

> In this section we describe how to use the Question Answerer on knowledge bases with long, free-form text or unstructured data. For example, a company's Human Resource (HR) team could have documents about general policies of the company. We can use this data to develop a HR assistant that automatically answers any policy related questions.

Reviewer: ```suggestion
In this section we describe how to use the question answerer on knowledge bases with long, free-form text or unstructured data. For example, a company's Human Resources (HR) team could have documents about general policies of the company. We can use this data to develop an HR assistant that automatically answers any policy-related questions.
```

> In this section we describe how to use the Question Answerer on knowledge bases with long, free-form text or unstructured data. For example, a company's Human Resource (HR) team could have documents about general policies of the company. We can use this data to develop a HR assistant that automatically answers any policy related questions.

Author: Seems like we put them in single quotes like 'text-to-speech', 'keyword spotting' etc

To use the question answerer on unstructured text, we would follow the same steps mentioned before to prepare, index and load the knowledge base.

> Here is an example of what the knowledge base can look like for the HR assistant usecase. It consists of frequently asked question and answer pairs.

Reviewer 2: ```suggestion

Here is an example of what the knowledge base can look like for the HR assistant use case. It consists of frequently asked question-answer pairs.
```

> Here is an example of what the knowledge base can look like for the HR assistant usecase. It consists of frequently asked question and answer pairs.

Reviewer: ```suggestion
Here is an example of what the knowledge base can look like for the HR assistant use case. It consists of frequently asked question and answer pairs.
```

.. code:: python

	from mindmeld.components import QuestionAnswerer
	qa = QuestionAnswerer(app_path='hr_assistant')
	qa.load_kb(app_namespace='hr_assistant', index_name='faq_data', data_file='hr_assistant/data/faq_data.json')

>   qa.get(index='faq_data')

Reviewer 2: ```suggestion
 qa.get(index='faq_data')
```

Reviewer: I think it should instead have an additional space to align with the previous lines.

.. code-block:: console

  [
      {
          "question": "What if I did not receive a form W2?",
          "answer": "W2s are mailed to home addresses of employees as of a date in mid-January each year. If you did not receive a form W2, you may access it online (beginning with 2008 W2s), similar to employee Pay Statements. Search for and click on the W-2 task. Enter the year for the W-2 you are wanting to locate and the last 4 digits of your Social Security Number. You then will have access to the W-2 in a format that can be used for filing paper versions of federal and state tax returns.",
          "id": "hrfaq6"
      },
      {
          "question": "Are employers expected to hire the less qualified over the more qualified to meet affirmative action goals?',
          "answer": "Employers are not expected to establish any hiring practices that conflict with the principles of sound personnel management. No one should be hired unless there is a basis for believing the individual is the best-qualified candidate. In fact, affirmative action calls for the hiring of qualified people.",
          "id": "hrfaq12"
      },
      ...
  ]

> The difference between the two cases is only in how MindMeld handles the search query. Internally, while the ranking algorithm remains the same for both cases, the features extracted for ranking is different and is optimized to handle long text passages rather than keyword phrases.

Reviewer 2: What are the two cases?

Reviewer: The difference between a structured knowledge base and an unstructured one is in how MindMeld handles the search query.

> The difference between the two cases is only in how MindMeld handles the search query. Internally, while the ranking algorithm remains the same for both cases, the features extracted for ranking is different and is optimized to handle long text passages rather than keyword phrases.

Reviewer: ```suggestion
The difference between the two cases is only in how MindMeld handles the search query. Internally, while the ranking algorithm remains the same for both cases, the features extracted for ranking are different and are optimized to handle long text passages rather than keyword phrases.
```

> The difference between the two cases is only in how MindMeld handles the search query. Internally, while the ranking algorithm remains the same for both cases, the features extracted for ranking is different and is optimized to handle long text passages rather than keyword phrases.

Reviewer 3: ```suggestion
The difference between the two cases is only in how MindMeld handles the search query. Internally, while the ranking algorithm remains the same for both cases, the features extracted for ranking are different and are optimized to handle long text passages rather than keyword phrases.
```

We can perform the same search using the :meth:`query()` API as well.

.. code:: python

	from mindmeld.components import QuestionAnswerer
	qa = QuestionAnswerer(app_path='hr_assistant')
	s = qa.build_search(index='faq_data')

	query = 'when do i get my w2 form'
	s.query(query_type='text', question=query, answer=query).execute()

.. code-block:: console

	[
      {
          'question': 'When can I expect my annual form W2?',
          'answer': 'All employee W2s are mailed on or around January 31 for the prior calendar year.',
          'id': 'hrfaq5'
      },
      {
          'question': 'What if I did not receive a form W2?',
          'answer': 'W2s are mailed to home addresses of employees as of a date in mid-January each year. If you did not receive a form W2, you may access it online (beginning with 2008 W2s), similar to employee Pay Statements. Search for and click on the W-2 task. Enter the year for the W-2 you are wanting to locate and the last 4 digits of your Social Security Number. You then will have access to the W-2 in a format that can be used for filing paper versions of federal and state tax returns.',
          'id': 'hrfaq6'
      }
      ...
  ]

.. note::

> 	For knowledge bases indexed prior to MindMeld 4.2, you will have to delete and reindex all the data to use this feature.

Reviewer: Would be good to provide commands for KB deletion and recreation.

Author: Need clarification as to what we want our recommended way to delete an index is. It seems like we do not expose  `delete_index()` to the user.

Reviewer 4: For entity resolution, the index is recreated when we use the parameter `clean=True` in the fit command. I think the cleanest solution would be to add a similar param to `load_kb`. When that parameter is true, it should delete and recreate the index.

Author: done

.. note::

   We can set the ``size`` parameter of the :meth:`execute()` method to specify the maximum number of records.

Dealing with unstructured text
------------------------------

> The knowledge bases described in the previous sections are comprised of clearly defined data types that are easily searchable. This kind of data is know as `structured data`.

Reviewer: ```suggestion
The knowledge bases described in the previous sections are comprised of clearly defined data types that are easily searchable. This kind of data is known as `structured data`.
```
